### ISSUE TITLE
Help Request - Dart examples

### ISSUE THREAD

**Issue by joeblew99** [swagger-api/swagger-codegen#4271]
I am using goalng on server and dart on the client

I see that Swagger now has Dart client code gen support:
https://github.com/swagger-api/swagger-codegen/tree/master/modules/swagger-codegen/src/main/resources/dart

Does anyone happen to know any examples of this used in projects or more code samples where this is used ?

thanks in advance. 
If there is a better place to ask this please let me know and i will close it here

**Comment by wing328** [comment]
@ircecho may I know if you've any sample codes/projects on using the Dart API client generated by Swagger Codegen?

It would be nice if the community can contribute the following: 
1) add some integration tests for Dart API client, 
2) add auto-generated documentation to Dart API client.

**Comment by ircecho** [comment]
Unfortunately, we do not have any publicly available code samples. And it would not be very interesting for this context either. Our API calls are pretty straightforward. For example:

``` dart
  defaultApiClient.basePath = 'http://localhost:8081/backend';
  final JwtHttpClient client = new JwtHttpClient();
  defaultApiClient.client = client;

  final String username = [...];
  final String password = [...];

  final LoginApi loginController = new LoginApi();
  final Credentials userToken = await loginController.login(new LoginData()
    ..email = username
    ..password = password);
  client.jwt = userToken.jwToken;
  Credentials merchantToken = await loginController.escalateToSellerAdmin(68);
  client.jwt = merchantToken.jwToken;
```
And so on. As you can see, we are doing nothing out of the ordinary.

**Comment by joeblew99** [comment]
thanks for response.

I think it would be really useful for others for examples that consume the output of the code gen to be in the repo.

So, shall i leave this open ?

**Comment by wing328** [comment]
@ircecho @joeblew99 if you guys have cycle, I can work with you to automatically generate the documentation for endpoints and models in the Dart API client, similar to what we've done for other languages: e.g. https://github.com/wing328/test-php/blob/master/docs/Api/PetApi.md#addpet

**Comment by ircecho** [comment]
@wing328 What would you need?

**Comment by wing328** [comment]
@ircecho as a starting point, I need to know which language (C#, Java, PHP, Ruby, etc) is similar to Dart so that we can borrow the documentation templates from other language to start with.

**Comment by wing328** [comment]
@ircecho here is the doc I've come up with: https://github.com/wing328/swagger-codegen/blob/dart_add_doc/samples/client/petstore/dart/swagger/README.md

Please review to see if it looks good from your perspective

Disclaimer: I'm no expert in Dart

**Comment by wing328** [comment]
PR: https://github.com/swagger-api/swagger-codegen/pull/4291

**Comment by ircecho** [comment]
The import statements need to be changed. Only the following is necessary:

```dart
import 'package:pet_store/api.dart';
```

The name `pet_store` depends on how the library is named, which is up to the user. You will know what is the correct value in this case.

Furthermore, it should be noted that it is necessary to include the library in the pubspec.yaml file. We are using a github reference and there are other ways, for instance including it from a local folder.

``` yaml
name: PetStoreExample
version: 0.0.1
description: A simple console application.
dependencies:
  pet_store:
    git: https://github.com/somethingsomething/pet_store.git
    version: 'any'
```

**Comment by wing328** [comment]
How's `pet_store` determined? Is it the `pubName`? I only found the following option for Dart client at the moment:
```
CONFIG OPTIONS
	sortParamsByRequiredFlag
	    Sort method arguments to place required parameters before optional parameters. (Default: true)

	ensureUniqueParams
	    Whether to ensure parameter names are unique in an operation (rename parameters that are not). (Default: true)

	browserClient
	    Is the client browser based

	pubName
	    Name in generated pubspec

	pubVersion
	    Version in generated pubspec

	pubDescription
	    Description in generated pubspec

	sourceFolder
	    source folder for generated code
```

> Furthermore, it should be noted that it is necessary to include the library in the pubspec.yaml file

I'll update the installation with that

Can you also show how to include it from a local folder as well?

**Comment by ircecho** [comment]
Yes. The three values have to be the same: `pubName`, the name used in `pubspec.yaml` and the name used in the import.

The pubspec would look like this:

```yaml
dependencies:
  pet_store:
    path: /Users/me/pet_store
```

A full treatise on dependencies can be found here: https://www.dartlang.org/tools/pub/dependencies Dart users will already have some superficial knowledge about that.

**Comment by wing328** [comment]
Updated: https://github.com/wing328/swagger-codegen/blob/dart_add_doc/samples/client/petstore/dart/swagger/README.md

Please take another look.

Can you also help review the sample code for authentication (HTTP basic, API key, OAuth)? https://github.com/wing328/swagger-codegen/blob/dart_add_doc/modules/swagger-codegen/src/main/resources/dart/api_doc.mustache#L31-L45

**Comment by ircecho** [comment]
Yes, this looks good now. Maybe you can add a hint, that `swagger` is dependent on the `pubName`.

As for the authentication, I cannot help you there. We do not use the built-in swagger authentication (even though we probably should :smile:).

**Comment by wing328** [comment]
> As for the authentication, I cannot help you there. We do not use the built-in swagger authentication (even though we probably should 😄).

That's ok. We'll leave those as "TODO" for now.

**Comment by wing328** [comment]
PR merged. 

@joeblew99 please pull the latest master to give it a try.

**Comment by joeblew99** [comment]
I will thanks

**Comment by wing328** [comment]
(If your company is using Swagger Codegen, please consider adding your company name to [this list](https://github.com/swagger-api/swagger-codegen#companiesprojects-using-swagger-codegen))

**Comment by ircecho** [comment]
@wing328 Sorry that I noticed this so late. :pensive: The imports in README.md have changed and are correct now, but the imports in the other pages are still as they were before and need also be changed. 

https://github.com/swagger-api/swagger-codegen/blob/master/samples/client/petstore/dart/swagger/docs/PetApi.md#addpet

**Comment by wing328** [comment]
@ircecho No worry :) and it will be fixed via https://github.com/swagger-api/swagger-codegen/pull/4292

Thanks for performing another check.

**Comment by wing328** [comment]
#4292 merged into master.

**Comment by wing328** [comment]
FYI. Dart API client already supports authentication (HTTP basic, API key, etc): https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/resources/dart/api_client.mustache#L24-L29

so we only need to update the documentation with the correct sample code so that developers can more easily consume the Dart API client.

cc @yissachar

**Comment by ircecho** [comment]
I assumed that the authentication was supported.

What I wanted to express was, that we are not using it in our clients yet even though it is supported and so I do not know how it works and cannot provide any examples.

**Comment by wing328** [comment]
@ircecho yes, I know what you mean. I just want others reading this thread aware that authentication is supported in Dart API client and they know where to look for the code by the link I provided.
